### PR TITLE
RECIRC 49: ABTest Fandom content in footer

### PR DESCRIPTION
**Pull request by pauloslund**
## Links

* https://wikia-inc.atlassian.net/browse/RECIRC-49
* Design: https://wikia-inc.atlassian.net/browse/RECIRC-50

## Description

Running an AB Test with Fandom content in the footer instead of the standard related pages. This is very similar to a test being run in Oasis with displaying Fandom content in the Right Rail. Given this is an AB Test, the code is considered temporary and will likely change in the near future

## Reviewers

@Wikia/x-wing

### CODE REVIEW

File: front/common/modules/AbTest.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+const AbTest = window.Wikia && window.Wikia.AbTest || {};
```
Comment (Warkot) [inline front/common/modules/AbTest.js]: You could probably just write `export default window…` as it's not a named export.

File: front/main/app/components/content-recommendations.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+			category: 'recirculation',
+			label: 'footer'
+		});
+	},
+
+	didRender() {
+		const pages = this.get('pages');
+
+		if (pages) {
+			Ember.run.throttle(this, 'trackImpression', 200);
```
Comment (Warkot) [inline front/main/app/components/content-recommendations.js]: Why do you need the throttle here? Isn't this hook run just once per page transition and rendering? It doesn't run for every recommended page, does it?

File: front/main/app/components/content-recommendations.js [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+		}
+	},
+
+	actions: {
+		trackClick() {
```
Comment (Warkot) [inline front/main/app/components/content-recommendations.js]: You could consider naming it a little bit different—like e.g. `trackExperimentClick`—to prevent any confusion with this mixin → https://github.com/Wikia/mercury/blob/RECIRC-49/front/main/app/mixins/track-click.js#L27.

File: front/main/app/templates/components/fandom-posts.hbs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+<section class="fandom-posts content-recommendations">
+	<h2>{{ title }}</h2>
```
Comment (Warkot) [inline front/main/app/templates/components/fandom-posts.hbs]: A minor thing—I don't think the spaces after `{{` and before `}}` are recommended.

**Comment by Warkot** [comment]
IMO, it looks pretty neat—especially for an experimental code. :+1:

File: front/main/app/components/content-recommendations.js (hunk 2 above)
Comment (pauloslund) [inline front/main/app/components/content-recommendations.js]: `didRender` was being fired twice in rapid succession when transitioning between articles. Not entirely sure why it was happening, my initial guess due to the article model potentially clearing the `relatedPages` property in between rendering the page but I could be entirely off base. I'll do some more research to figure out the reason.

**Comment by vforge** [comment]
@pauloslund I have more of a general question: Is there any reason not ot use optimizely for this? It seems a bit redundant to implement our in-house AbTesting when we have working solution already. 
Also I think we should consider sending proper dimensions (that descibes our in-house AbTesting tests and current groups) to the UA as we currently are doing in Oasis/Desktop.

**Comment by pauloslund** [comment]
@vforge The primary reasons for not using Optimizely for these experiments are because of Optimizely's usage limits. We're limited on the number of impressions as well as the number of experiments we can run that report to GA (custom dimensions 48-50 are reserved for Optimizely with 48, and 49 already being in use for other experiments). This is only the initial set of experiments we plan to run for recirculation on Mercury so we felt it made sense to leverage the AbTesting tool.

Very good point about sending proper dimensions to the UA.

**Comment by vforge** [comment]
@pauloslund Thanks for clarification! :+1:
Do you have plan to remove all testing code from Mercury and AbTesting with it? Or we want to leave it in our codebase? (I'm just curious)

File: front/common/modules/AbTest.js [hunk 5]
```diff
@@ -0,0 +1 @@
+export default window.Wikia && window.Wikia.AbTest || {};
```
Comment (rogatty) [inline front/common/modules/AbTest.js]: If you return an empty object then `AbTest.getGroup(experiment)` will fail with `getGroup is not a function` or something similar.

File: front/main/app/components/ab-test.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
+import Ember from 'ember';
+import AbTest from 'common/modules/AbTest';
+
+export default Ember.Component.extend({
+	willRender() {
```
Comment (rogatty) [inline front/main/app/components/ab-test.js]: What we do in this hook isn't really related to rendering stuff, right? How about using `didReceiveAttrs()`?
https://guides.emberjs.com/v2.4.0/components/the-component-lifecycle

Also, please call `this._super(...arguments)` when overriding Ember hooks.

Comment (rogatty) [inline front/main/app/components/ab-test.js]: What we do in this hook isn't really related to rendering stuff, right? How about using `didReceiveAttrs()`?
https://guides.emberjs.com/v2.4.0/components/the-component-lifecycle

Also, please call `this._super(...arguments)` when overriding Ember hooks.

File: front/main/app/components/content-recommendations.js (hunk 2 above)
Comment (rogatty) [inline front/main/app/components/content-recommendations.js]: You can use `didReceiveAttrs()` instead of `didRender()`. It's called only once per article view.

File: front/main/app/components/fandom-posts.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+			}, (error) => {
+				// At some point this error should possibly be handled
+			});
+	},
+
+	didRender() {
+		const posts = this.get('posts');
+
+		if (posts) {
+			Ember.run.throttle(this, 'trackImpression', 200);
```
Comment (rogatty) [inline front/main/app/components/fandom-posts.js]: The same as https://github.com/Wikia/mercury/pull/2081/files?diff=split#r54878608

File: front/main/app/models/fandom-posts.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
+import Ember from 'ember';
+
+const FandomPostsModel = Ember.Object.extend({
+	title: null,
+	type: null,
+	posts: [],
```
Comment (rogatty) [inline front/main/app/models/fandom-posts.js]: It's risky, see https://wikia-inc.atlassian.net/wiki/pages/viewpage.action?pageId=44007507

File: front/common/modules/AbTest.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+	}
+
+	const abList = AbTest.getExperiments(true);
+
+	for (let abIndex = 0; abIndex < abList.length; abIndex++) {
+		const abExp = abList[abIndex],
+			abSlot = AbTest.getGASlot(abExp.name);
+
+		if (abExp && abExp.flags && abExp.flags.ga_tracking) {
+			if (abSlot >= 40 && abSlot <= 49) {
```
Comment (hakubo) [inline front/common/modules/AbTest.js]: what is 40 and 49?
can you add a comment about that?

File: front/main/app/components/fandom-posts.js [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+	init() {
+		this._super(...arguments);
+
+		FandomPostsModel
+			.create({
+				type: this.get('type')
+			})
+			.load()
+			.then((model) => {
```
Comment (hakubo) [inline front/main/app/components/fandom-posts.js]: hmm this looks odd, 
what do you thing about setting
this.set('model', Fandom...)

and then access the model data through `model.`

this would be more clear where the data is coming from

File: front/main/app/templates/components/article-wrapper.hbs [hunk 11]
```diff
@@ -111,7 +111,15 @@
 		{{/collapsible-menu}}
 	{{/if}}
 </section>
-{{content-recommendations pages=model.relatedPages}}
+{{#ab-test experiment='RECIRCULATION_MERCURY_FOOTER' as |group| }}
```
Comment (hakubo) [inline front/main/app/templates/components/article-wrapper.hbs]: neat :)

File: front/main/app/templates/components/fandom-posts.hbs [hunk 12]
```diff
@@ -0,0 +1,17 @@
+{{#if posts.length}}
+<section class="fandom-posts content-recommendations">
+	<h2>{{title}}</h2>
+	<ul class="thumbnails">
+		{{#each posts as |post|}}
+			<li class="title-thumbnail" {{action 'trackExperimentClick' preventDefault=false}}>
+				<a href="{{post.url}}" class="image image-thumbnail">
+					<img src="{{post.image_url}}" alt="{{post.title}}">
+				</a>
+				<p class="title" title="{{post.title}}">
```
Comment (hakubo) [inline front/main/app/templates/components/fandom-posts.hbs]: never saw someone setting title for p, what is it for? tooltip?

also you have 2 links next to each other that point to the same place
could this a wrap the whole content - so the whole thing is clickable?

File: front/main/app/components/fandom-posts.js [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+	init() {
+		this._super(...arguments);
+
+		const model = FandomPostsModel.create({
+			type: this.get('type')
+		});
+
+		model.load();
+		this.set('model', model);
```
Comment (hakubo) [inline front/main/app/components/fandom-posts.js]: there are no listeners set up when in init hook
so you could actually do
this.model = model;

**Comment by pauloslund** [comment]
Got a final LGTM from @hakubo over Slack so I'm going to merge this branch

File: front/common/modules/AbTest.js [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+	const abList = AbTest.getExperiments(true);
+
+	for (let abIndex = 0; abIndex < abList.length; abIndex++) {
+		const abExp = abList[abIndex],
+			abSlot = AbTest.getGASlot(abExp.name);
+
+		if (abExp && abExp.flags && abExp.flags.ga_tracking) {
+			// GA Slots 40-49 are reserved for our AB Testing tool. Anything outside that
+			// range could potentially overwrite something that we don't want to
+			if (abSlot >= 40 && abSlot <= 49) {
```
Comment (rogatty) [inline front/common/modules/AbTest.js]: Are you sure about these numbers? According to https://one.wikia-inc.com/wiki/Engineering/ABTesting/Optimizely#Google_Analytics_Integrations dimensions 48-50 are used for Optimizely.
/cc @Warkot @vforge

File: front/common/modules/AbTest.js [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+	}
+}
+
+/**
+ * Integrates AbTest with Universal Analytics
+ *
+ * @param {Array} dimensions
+ * @returns {Array}
+ */
+export function integrateAbTestWithUA(dimensions) {
```
Comment (rogatty) [inline front/common/modules/AbTest.js]: This method could really use some unit test :) I'm fine with merging this PR without it and adding them later, though.

File: front/common/modules/AbTest.js (hunk 14 above)
Comment (pauloslund) [inline front/common/modules/AbTest.js]: Hmm, interesting. I should have mentioned this in the code but this is pulled almost directly from the code in Oasis. There is actually a test being run in Oasis currently that is tracking on slot 48. Not entirely sure what the best way to handle this would be. We can limit it down to 40-47 on Mercury however that puts it out of sync with the main site

File: front/common/modules/AbTest.js (hunk 15 above)
Comment (pauloslund) [inline front/common/modules/AbTest.js]: Unit tests will be following shortly in another PR :)

File: front/common/modules/AbTest.js (hunk 14 above)
Comment (Warkot) [inline front/common/modules/AbTest.js]: There's a Slack channel #optimizely, you raise this topic there.

Comment (rogatty) [inline front/common/modules/AbTest.js]: If it's the same issue in Oasis then IMO this can be merged and resolved later, globally.

**Comment by rogatty** [comment]
LGTM

File: front/common/modules/AbTest.js (hunk 14 above)
Comment (pauloslund) [inline front/common/modules/AbTest.js]: Sounds like a plan

Comment (vforge) [inline front/common/modules/AbTest.js]: @pauloslund @rogatty You're both right - I remember that we've selected dimensions for Optimizely not to collide with in-house AbTesting - I couldn't find original document about it, though.
I suggest editing this page on WikiaOne and put information that we can use dimension >= 50 in Optimizely - followed by an e-mail.
Alternatively you can ask Zach Hobbs, as it's he who added this information to One - https://one.wikia-inc.com/index.php?title=Engineering%2FABTesting%2FOptimizely&diff=248985&oldid=248696
